### PR TITLE
Support of SQLCMD variables in Project & Package references

### PR DESCRIPTION
**Pull request by McHarny**
Standard SSDT projects has ability to reference other database not only by `DatabaseVariableLiteralValue` but also by using `SQLCMD` variables to use in code for reference another database/server. Like this:
```xml
<ProjectReference Include="SomeProject.sqlproj">
      <Name>SomeProject</Name>
      <Project>{c8cef164-7da7-4fb6-9fcf-41bcd3257218}</Project>
      <Private>True</Private>
      <SuppressMissingDependenciesErrors>False</SuppressMissingDependenciesErrors>
      <DatabaseSqlCmdVariable>SomeDatabase</DatabaseSqlCmdVariable>
      <ServerSqlCmdVariable>SomeServer</ServerSqlCmdVariable>
</ProjectReference>
...
<ItemGroup>
  <SqlCmdVariable Include="SomeDatabase">
      <DefaultValue>SomeDatabase</DefaultValue>
      <Value>$(SqlCmdVar__1)</Value>
    </SqlCmdVariable>
    <SqlCmdVariable Include="SomeServer">
      <DefaultValue>SomeServer</DefaultValue>
      <Value>$(SqlCmdVar__2)</Value>
  </SqlCmdVariable>
</ItemGroup>
```
And objects from this project can be reference by `[$(SomeServer)].[$(SomeDatabase)].[<schema>].[<object>]` syntax.

I add similar functionality to SqlProj.
Since the functionality of `DatabaseVariableLiteralValue` is based on attributes, not elements as in `.sqlproj`, I added two new attributes to the `PackageReference` and `ProjectReference` elements with the names similar to `.sqlproj`:
```
DatabaseSqlCmdVariable
ServerSqlCmdVariable
```
Their behavior is completely similar to that of `.sqlproj`. You can specify them in the reference, define appropriate `SQLCMD` variables and use reference in code. You can combine them with `DatabaseVariableLiteralValue` for example:
```xml
<Project Sdk="MSBuild.Sdk.SqlProj/1.12.0">
    <PropertyGroup>
        <TargetFramework>netstandard2.0</TargetFramework>
    </PropertyGroup>

    <ItemGroup>
        <ProjectReference Include="../MyOtherProject/MyOtherProject.csproj" DatabaseVariableLiteralValue="SomeOtherDatabase" />
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="../MySecondProject/MySecondProject.csproj" DatabaseSqlCmdVariable="SecondOtherDatabase" ServerSqlCmdVariable="SomeOtherServer" />
    </ItemGroup>
    <ItemGroup>
        <ProjectReference Include="../MyThirdProject/MyThirdProject.csproj" DatabaseVariableLiteralValue="ThirdOtherDatabase" ServerSqlCmdVariable="SomeOtherServer" />
    </ItemGroup>

  <ItemGroup>
    <SqlCmdVariable Include="SecondOtherDatabase">
      <DefaultValue>SecondDatabase</DefaultValue>
      <Value>$(SqlCmdVar__1)</Value>
    </SqlCmdVariable>
    <SqlCmdVariable Include="SomeOtherServer">
      <DefaultValue>OtherServer</DefaultValue>
      <Value>$(SqlCmdVar__2)</Value>
    </SqlCmdVariable>
  </ItemGroup>

</Project>
```
For this example we can reference other objects with those combinations:
```
[SomeOtherDatabase].[<schema>].[<object>]
[$(SomeOtherServer)].[$(SecondOtherDatabase)].[<schema>].[<object>]
[$(SomeOtherServer)].[ThirdOtherDatabase].[<schema>].[<object>]
```

To do this, I changed the format of the `--reference` (`-r`) parameter. Previously, it took the value `<Path>;<DatabaseVariableLiteralValue>`, now it takes the value `<Path>;<External parts>`, where `<External parts>` has the format `dbl=<DatabaseVariableLiteralValue>|dbv=<DatabaseSqlCmdVariable>|srv=<ServerSqlCmdVariable>`

In `Extensions` class I add method `ParseExternalParts` which parses this `<External parts>`. Parsing of this is made similar to that is writed in constructor of `Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlSchemaCustomSchemaDataWrapper`

Also parsing supports old behavior with simple `<DatabaseVariableLiteralValue>` passing in the parameter.

I add some Unit-tests to `ExtensionsTest` and add test project `TestProjectWithExternalServerReference`

In addition I slightly changed the behavior of the warning processing to make it more similar to the one in `.sqlproj`. Now the message of warnings threated as error wrote as `error`. Suppressed warnings are not appear in output. 
I remove `ModelValidationError` clause from output text for compatibility with some VS extensions which parses output and colorize it based on regex.

I also correct `CanGetModelValidationErrors` test bacause it failed on localized versions of `.NET`.
Also I have problems with `CanParseFileWithIncludesFromRelativePaths` test on some machines with mixing of `\` and `/` in relative paths, so I add  replacing `\` to `/` in target. It doesn't affect the functionality of test

### CODE REVIEW

File: src/DacpacTool/Extensions.cs [hunk 1]
```diff
@@ -107,26 +106,101 @@ public static void AddReference(this TSqlModel model, string referencePath, stri
...
+                if (!string.IsNullOrEmpty(parts))
+                {
+                    setMetadataMethod.Invoke(customData, new object[] {"ExternalParts", parts});
+                }
             }
 
             AddCustomData(dataSchemaModel, customData);
         }
 
+        public static string ParseExternalParts(string externalParts)
```
Comment (jmezach) [inline src/DacpacTool/Extensions.cs:119]: I'm wondering if perhaps a regular expression would be a better solution for this parsing logic. It feels fairly complicated.

File: src/DacpacTool/ModelValidationError.cs [hunk 2]
```diff
@@ -72,5 +72,30 @@ public override string ToString()
 
             return stringBuilder.ToString();
         }
+
+        /// <summary>
+        /// Make message for writing to output during build process more similar to the one displayed by standard SSDT build
+        /// </summary>
+        /// <param name="severity">Severity to override depends on TreatTSqlWarningsAsErrors</param>
+        /// <returns></returns>
+        public string GetOutputMessage(ModelErrorSeverity severity)
```
Comment (jmezach) [inline src/DacpacTool/ModelValidationError.cs:81]: The original goal of the override of `ToString()` was to do exactly this. Perhaps that override is no longer necessary if we add this?

File: src/DacpacTool/PackageBuilder.cs [hunk 3]
```diff
@@ -112,19 +112,26 @@ public bool ValidateModel()
                 if (modelError.Severity == ModelErrorSeverity.Error)
                 {
                     validationErrors++;
+                    Console.WriteLine(modelError.GetOutputMessage(modelError.Severity));
                 }
-                else if (modelError.Severity == ModelErrorSeverity.Warning && TreatTSqlWarningsAsErrors)
+                else if (modelError.Severity == ModelErrorSeverity.Warning)
                 {
                     if (!_suppressedWarnings.Contains(modelError.ErrorCode))
```
Comment (jmezach) [inline src/DacpacTool/PackageBuilder.cs:119]: This is getting a little messy with multiple nested if's. Perhaps we could fold the first two if's here into a single method call.

File: test/DacpacTool.Tests/ExtensionsTest.cs [hunk 4]
```diff
@@ -73,7 +73,121 @@ public void CanGetModelValidationErrors()
 
             var error = modelValidationErrors.First();
             error.Severity.ShouldBe(Microsoft.SqlServer.Dac.Model.ModelErrorSeverity.Error);
-            error.ToString().ShouldBe("../../../../TestProjectWithErrors/Procedures/csp_Test.sql(2,18):ModelValidationError Error SQL71501: SqlSubroutineParameter: [dbo].[csp_Test].[@p_Parameter] has an unresolved reference to SqlBuiltInType [dbo].[MyCustomType].");
+            error.ErrorCode.ShouldBe(71501);
```
Comment (jmezach) [inline test/DacpacTool.Tests/ExtensionsTest.cs:76]: Did we remove the assertion on ToString() for a reason? I think that still has value to avoid regressions in the output we give.

File: test/DacpacTool.Tests/ScriptParserTests.cs [hunk 5]
```diff
@@ -62,7 +62,7 @@ public void CanParseFileWithIncludesFromRelativePaths()
             // Assert
             result.ShouldNotBeNull();
             result.Count().ShouldBe(2);
-            result.ToList().Take(2).Last().ShouldBe($"{TEST_PROJECT_PATH}/Pre-Deployment/./MoreScripts/Script2.sql");
+            result.ToList().Take(2).Last().Replace("\\", "/").ShouldBe($"{TEST_PROJECT_PATH}/Pre-Deployment/./MoreScripts/Script2.sql");
```
Comment (jmezach) [inline test/DacpacTool.Tests/ScriptParserTests.cs:65]: Was this change deliberate? This appears to be a platform specific difference, so perhaps we should handle this better?

File: test/TestProjectWithExternalServerReference/TestProjectWithExternalServerReference.csproj [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <SqlServerVersion>Sql150</SqlServerVersion>
+  </PropertyGroup>
+  
+  <ItemGroup>
+    <ProjectReference Include="..\TestProject\TestProject.csproj" DatabaseVariableLiteralValue="SomeDatabase" ServerSqlCmdVariable="SomeServer"/>
+  </ItemGroup>
+
+  <ItemGroup>
+    <SqlCmdVariable Include="SomeDatabase">
```
Comment (jmezach) [inline test/TestProjectWithExternalServerReference/TestProjectWithExternalServerReference.csproj:14]: This variable isn't actually needed right? Perhaps we should remove it then.

**Review by jmezach — CHANGES_REQUESTED** [review]
Once again, thank you for your contribution @McHarny. Overall I think this looks good, just a couple of comments.

File: src/DacpacTool/ModelValidationError.cs (hunk 2 above)
Comment (McHarny) [reply src/DacpacTool/ModelValidationError.cs:81]: IMHO override of `ToString` is good practice for develop & debug purpose.

File: test/DacpacTool.Tests/ExtensionsTest.cs (hunk 4 above)
Comment (McHarny) [reply test/DacpacTool.Tests/ExtensionsTest.cs:76]: If we has localized `.NET` on machine running tests it report error message other than that with which you are compare. So I think that we must check `ModelValidationError` properties to check correct result, instead of check `ToString`

File: test/DacpacTool.Tests/ScriptParserTests.cs (hunk 5 above)
Comment (McHarny) [reply test/DacpacTool.Tests/ScriptParserTests.cs:65]: OK. This issue depends on platform that running tests. Windows use `\`, *nix use `/`, Some other magic platform might use `-` for example. IMHO comparing static string with OS results is not a good idea, So I think that it is good place to make refactoring. But for now I only correct line to avoid failing on main platforms (*nix & windows). Editing all unit-tests is not a deal of this PR. I only make editions for tests don't failure. I think that you can do proper changes after that PR.

File: src/DacpacTool/PackageBuilder.cs (hunk 3 above)
Comment (McHarny) [reply src/DacpacTool/PackageBuilder.cs:119]: To `if..else` statement we can switch to `switch` statement, it will reduce if'ing on top level.

**Comment by ErikEJ** [comment]
Add timeout to RegEx ...

**Comment by McHarny** [comment]
@jmezach will be any progress?

**Comment by jmezach** [comment]
It looks like CI is failing again and I think I've figured out why. We're using some source code from https://github.com/microsoft/sqltoolsservice and the commit we're using uses a NuGet feed that is no longer available. I think we'll need to update the submodule to use the latest commit on `main` from that repo. @McHarny Can you make that change so that CI will succeed?

**Comment by McHarny** [comment]
@jmezach ok. I will try

**Comment by McHarny** [comment]
@jmezach I see to the `sqltoolsservice` project. Invalid NuGet feed is written in nuget.config, that presents in the latest version of this submodule too. the wlast change of this file is dated by 2019 year.
I think that It need to edit CI definition to not to build this submodule.

**Review by jeffrosenberg — CHANGES_REQUESTED** [review]

File: src/DacpacTool/Extensions.cs [hunk 7]
```diff
@@ -107,26 +107,92 @@ public static void AddReference(this TSqlModel model, string referencePath, stri
...
+            {
+                if(match.Groups["dbl"].Success)
+                    databaseVariableLiteralValue = match.Groups["dbl"].Value;
+                else if (match.Groups["dbv"].Success)
+                    databaseVariableName = match.Groups["dbv"].Value;
+                else if (match.Groups["srv"].Success)
+                    serverVariableName = match.Groups["srv"].Value;
+            }
+
+            if (string.IsNullOrEmpty(serverVariableName) && string.IsNullOrEmpty(databaseVariableName) &&
```
[the remark this answers is not in the document]
Comment (McHarny) [reply src/DacpacTool/Extensions.cs:136]: @jeffrosenberg  Codebase of this is fully taken from `Microsoft.Data.Tools.Schema.Sql.SchemaModel.SqlSchemaCustomSchemaDataWrapper` constructor. If you want's me to refactor this method I can do that. But in generally there are four working combination of these three variables, so four `if`'s will be always present. IMHO

File: test/DacpacTool.Tests/ScriptParserTests.cs (hunk 5 above)
Comment (McHarny) [reply test/DacpacTool.Tests/ScriptParserTests.cs:65]: I have Windows 10 Pro 19042.928 installed on my machine. With default Russian locale and Russian language. Also I have language pack installed to .NET SDK. I don't know this is the reason of error or not but the result of test run on picture at attach
Debug

File: src/DacpacTool/Extensions.cs [hunk 8]
```diff
@@ -107,26 +108,90 @@ public static void AddReference(this TSqlModel model, string referencePath, stri
...
+            }
+            else
+            {
+                databaseVariableLiteralValue = Identifier.EncodeIdentifier(externalParts);
+            }
+
+            // if there are no value for database return. Reference MUST have database value
+            if (string.IsNullOrEmpty(databaseVariableLiteralValue) && string.IsNullOrEmpty(databaseVariableName))
+                return null;
+
```
Comment (ErikEJ) [inline src/DacpacTool/Extensions.cs:156]: nit: Remove empty line

File: src/DacpacTool/PackageBuilder.cs [hunk 9]
```diff
@@ -112,19 +112,16 @@ public bool ValidateModel()
                 if (modelError.Severity == ModelErrorSeverity.Error)
                 {
                     validationErrors++;
+                    Console.WriteLine(modelError.GetOutputMessage(modelError.Severity));
```
Comment (ErikEJ) [inline src/DacpacTool/PackageBuilder.cs:115]: Why is this part of this PR?

File: test/DacpacTool.Tests/TestModelBuilder.cs [hunk 10]
```diff
@@ -41,9 +41,15 @@ public TestModelBuilder AddStoredProcedureFromFile(string filename)
             return this;
         }
 
-        public TestModelBuilder AddReference(string path)
+        public TestModelBuilder AddReference(string path, string externalParts = "")
```
Comment (ErikEJ) [inline test/DacpacTool.Tests/TestModelBuilder.cs:44]: "" => string.Empty

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: test/DacpacTool.Tests/TestModelBuilder.cs (hunk 10 above)
Comment (McHarny) [reply test/DacpacTool.Tests/TestModelBuilder.cs:44]: `string.Empty` doesn't allow to use in parameters default.

File: src/DacpacTool/PackageBuilder.cs (hunk 9 above)
Comment (McHarny) [reply src/DacpacTool/PackageBuilder.cs:115]: Because of I and my team actively use this great project in our development. Using NuGet packages is what really amazing feature. But we need some features witch presence in SSDT projects and are absent in SqlProj. So I decide to add those features to SqlProj. But approving of them is too slow. All of my additions was developed in two days. As @jmezach ask to split them to partials PR. I do It. but every approve take more than 2 weeks to proceed..... So I think, that It will be more productive to us to maintain our fork separately from the main project. Sorry, I may be newbie for git and GitHub, I've using tfvc for many years, and vss before it. And I began to learn git about some months ago. So switching between branches, rebasing, pull requesting and so, and so, is too hard for me (for now).

**Review by ErikEJ — APPROVED** [review]

**Comment by McHarny** [comment]
@jmezach , @jeffrosenberg can you approve?

**Review by jeffrosenberg — APPROVED** [review]

**Review by jmezach — APPROVED** [review]
LGTM

**Comment by jmezach** [comment]
@McHarny Version 1.14.0 is up on NuGet now which includes these changes.
